pkg/reconciler/delivery/policy.go

```diff
@@ -52,29 +54,65 @@ type Stage struct {
 // computeNewPercent calculates, given a Policy and the current rollout stage,
 // the traffic percentage for the NEW Revision in the next rollout stage
 func computeNewPercent(p *Policy, currentPercent int) (int, error) {
-	i := sort.Search(len(p.Percents), func(i int) bool {
-		return p.Percents[i].Percent >= currentPercent
+	i := sort.Search(len(p.Stages), func(i int) bool {
+		return p.Stages[i].Percent >= currentPercent
 	})
-	if i < len(p.Percents) && p.Percents[i].Percent == currentPercent {
-		if i == len(p.Percents) - 1 {
+	if i < len(p.Stages) && p.Stages[i].Percent == currentPercent {
+		if i == len(p.Stages) - 1 {
 			return 100, nil
 		}
-		return p.Percents[i + 1].Percent, nil
+		return p.Stages[i + 1].Percent, nil
 	}
 	return 0, fmt.Errorf("invalid percentage for current rollout stage")
 }
 
 // getThreshold returns, given the percentage for a rollout stage, its corresponding threshold value
 // if the threshold value isn't specified, DefaultThreshold is used
 func getThreshold(p *Policy, currentPercent int) (int, error) {
-	i := sort.Search(len(p.Percents), func(i int) bool {
-		return p.Percents[i].Percent >= currentPercent
+	i := sort.Search(len(p.Stages), func(i int) bool {
+		return p.Stages[i].Percent >= currentPercent
 	})
-	if i < len(p.Percents) && p.Percents[i].Percent == currentPercent {
-		if p.Percents[i].Threshold != nil {
-			return *p.Percents[i].Threshold, nil
+	if i < len(p.Stages) && p.Stages[i].Percent == currentPercent {
+		if p.Stages[i].Threshold != nil {
+			return *p.Stages[i].Threshold, nil
 		}
 		return p.DefaultThreshold, nil
 	}
 	return 0, fmt.Errorf("invalid percentage for current rollout stage")
 }
+
+// computeNewPercentExplicit is an explicit way of computing a percentage without relying on the previous stage
+// elapsed is the total time duration since the beginning of the rollout
+// this function doesn't return an error because an error is impossible
+func computeNewPercentExplicit(p *Policy, elapsed time.Duration) int {
+	metric := float64(elapsed) / float64(time.Second)
+	metricCumulative := 0
+	for _, s := range p.Stages[1:] {
```

Comment:
if stages has just 1 element? Or none?

Reply:
Yes, you are definitely right!!! I haven't written unit tests for this function, so this is what happens when I don't write unit tests :)

Reply:
I fixed this for the case of 0 elements. For 1 element, the for loop will be skipped, so the function still returns the desired result at the bottom.